### PR TITLE
Add AliasSeqOf to std.meta

### PR DESCRIPTION
**Pull request by jcrapuchettes**
Add template that allows for conversion of arrays and input ranges to AliasSeqs.

### CODE REVIEW

**Comment by Hackerpilot** [comment]
:white_check_mark:

**Comment by Hackerpilot** [comment]
Context:
https://github.com/D-Programming-Language/phobos/pull/3369
https://github.com/D-Programming-Language/phobos/pull/1472

**Comment by JakobOvrum** [comment]
PR #1472 had an optimization that yields less template instantiations for arrays, which is something that actually can be applied to all ranges with length and slicing. Would this be worth adding?

File: std/meta.d [hunk 1]
```diff
@@ -863,6 +863,42 @@ unittest
...
+    import std.string : capitalize;
+
+    struct S
+    {
+        int a;
+        int c;
+        int b;
+    }
+
+    alias capMembers = AliasSeqOf!([__traits(allMembers, S)].sort().map!capitalize().array());
```
Comment (atilaneves) [inline std/meta.d]: Why the call to `array` if `AliasSeqOf` takes any input range?

Comment (jcrapuchettes) [inline std/meta.d]: Good question and I've now removed it.

**Comment by jcrapuchettes** [comment]
After evaluating the initial implementation and the implementation from #1472, I found that the #1472 was significantly better. It was able to handle ranges and arrays greater than 250 elements. To that end, I've copied the implementation from #1472, modified it a little, and copied in its unit tests.

Thanks to @monarchdodra!

File: std/meta.d [hunk 2]
```diff
@@ -863,6 +863,97 @@ unittest
...
+    static assert(capMembers[0] == "A");
+    static assert(capMembers[1] == "B");
+    static assert(capMembers[2] == "C");
+}
+
+///
+unittest
+{
+    enum REF = [0, 1, 2, 3];
+    foreach(I, V; aliasSeqOf!([0, 1, 2, 3]))
```
Comment (JakobOvrum) [inline std/meta.d]: Nitpick: `I` and `V` will never refer to types, so they should not be capitalized.

**Comment by JakobOvrum** [comment]
LGTM. I would have preferred `toAliasSeq` but I'm fine with the current name too.

**Comment by Hackerpilot** [comment]
Auto-merge toggled on

**Comment by monarchdodra** [comment]
Awesome!

I guess I don't get a say in this anymore, but if I had to write my original implementation again, I'd only implement support for arrays. `std.meta` should not have a dependency on `std.range`, as `range` is more of a library concept than language concept (and `meta` is more of a language support module). That and having the caller do `aliasSeqOf!(myRange.array())` is trivially easy.

I'm not sure how much we are still trying to untangle library dependencies?

**Comment by JakobOvrum** [comment]
@monarchdodra, good point, I agree it's better to leave the conversion to array to the caller. Maybe file a PR? I don't think this is part of the new release.

**Pull request merged**